Source/DafnyCore/Auditor/AuditReport.cs

```diff
@@ -0,0 +1,154 @@
+  private Dictionary<Declaration, AssumptionProperties> assumptionsByDecl = new();
+
+  private void UseDecl(Declaration decl) {
+    declsWithEntries.Add(decl);
+    switch (decl) {
+      case MemberDecl memberDecl:
+        UseDecl(memberDecl.EnclosingClass);
+        break;
+      case TopLevelDeclWithMembers topDecl:
+        modulesWithEntries.Add(topDecl.EnclosingModuleDefinition);
+        break;
+    }
+  }
+  public void AddAssumptions(Declaration decl, AssumptionProperties assumptions) {
+    if (assumptions.HasAssumptions()) {
+      assumptionsByDecl.Add(decl, assumptions);
+      UseDecl(decl);
+    }
+  }
+
+  public IEnumerable<Assumption> AllAssumptions() {
+    return assumptionsByDecl.Select((e, _) => new Assumption(e.Key, e.Value));
+  }
+
+  private string RenderRow(string beg, string sep, string end, IEnumerable<string> cells) {
+    return beg + String.Join(sep, cells) + end;
+  }
+
+  private string GetFullName(Declaration decl) {
+    if (decl is MemberDecl m) {
+      return m.FullDafnyName;
+    } else if (decl is ModuleDecl mod) {
+      return mod.FullDafnyName;
+    } else if (decl is TopLevelDecl tld) {
+      return tld.FullDafnyName;
+    } else {
+      return decl.Name;
+    }
+  }
+  private IEnumerable<String> IssueRow(Assumption a, string issue, string mitigation) {
+    return new List<string>() {
+      GetFullName(a.decl),
+      (!a.props.IsGhost).ToString(),
+      a.props.HasAxiomAttribute.ToString(),
+      a.props.HasExternAttribute.ToString(),
+      issue,
+      mitigation
+    };
+  }
+
+  private string RenderAssumptionRows(Assumption a, string beg, string sep, string end) {
+    var rows = a.props
+      .Description()
+      .Select((desc, _) => RenderRow(beg, sep, end, IssueRow(a, desc.Item1, desc.Item2)));
+    return rows.Count() > 0 ? String.Join("\n", rows) : a.ToString();
+  }
+
+  private string RenderAssumptionRowsMarkdown(Assumption a) {
+    return RenderAssumptionRows(a, "| ", " | ", " |");
+  }
+
+  private string RenderAssumptionRowsHTML(Assumption a) {
+    return RenderAssumptionRows(a, "<tr><td>", "</td><td>", "</td></tr>");
+  }
+
+  public string RenderHTMLTable() {
+    var header =
+      "<tr><th>Name</th><th>Compiled</th><th>Explicit Assumption</th>" +
+      "<th>Extern</th><th>Issue</th><th>Mitigation</th></tr>\n";
+    var rows = AllAssumptions().Select(RenderAssumptionRowsHTML);
+    return header + String.Join("\n", rows);
+  }
+
+  public string RenderMarkdownTable() {
+    var header =
+      "|Name|Compiled|Explicit Assumption|Extern|Issue|Mitigation|\n" +
+      "|----|--------|-------------------|------|-----|----------|\n";
+    var rows = AllAssumptions().Select(RenderAssumptionRowsMarkdown);
+    return header + String.Join("\n", rows);
+  }
+
+  public string RenderMarkdownSections() {
+    StringBuilder text = new StringBuilder();
+
+    foreach (var module in modulesWithEntries) {
+      text.AppendLine($"# Module `{module.Name}`");
```

Comment:
Would it make sense to make the audit smaller by not reporting types and members if there is no audit warning in that module anyway? To achieve that, you could store the temporary strings and not emit sub-sections if they are empty. That way, the report would only contain the assumptions.

I'm not using the auditor myself so I don't know what is really useful for Dafny users. You probably know better.

Reply:
If I understand what you're saying, that's already what it's doing. The `modulesWithEntries` collection includes only the modules that have audit warnings.

Reply:
Does the filtering works as well on top-level declarations, and members? That's also something I had in mind

Reply:
Yes, see lines 121 and 135.